src/chunkserver/scan_manager.cpp

```diff
@@ -21,156 +21,367 @@
+        waitInterval_.WaitForNextExcution();
     }
-
-    return 0;
+    LOG(INFO) << "Scan worker thread stopped.";
 }
 
-void ScanManager::Enqueue(LogicPoolID poolId, CopysetID id) {
-    WriteLockGuard writeGuard(rwLock_);
-    ScanKey key(poolId, id);
-    if (waitScanSet.find(key) == waitScanSet.end()) {
-        waitScanSet.insert(key);
+void ScanManager::StartScanJob(ScanKey key) {
+    std::shared_ptr<ScanJob> job = std::make_shared<ScanJob>();
+    job->poolId = key.first;
+    job->id = key.second;
+    job->type = ScanType::Init;
+    job->currentChunkId = 0;
+    job->startTime = 0;
+    // check copyset exist
+    auto nodePtr = copysetNodeManager_->GetCopysetNode(job->poolId, job->id);
+    if (nullptr == nodePtr) {
+        LOG(ERROR) << "scan copyset failed, copyset node is not found: "
+                   << job->poolId << "," << job->id;
+        return;
     }
-    return;
-}
-
-ScanKey ScanManager::Dequeue() {
-    ScanKey key;
-    WriteLockGuard writeGuard(rwLock_);
-    key = *(waitScanSet.begin());
-    waitScanSet.erase(waitScanSet.begin());
-    return key;
+    dataStore_ = nodePtr->GetDataStore();
+    nodePtr->SetScan(true);
+    jobMapLock_.WRLock();
+    jobs_.emplace(key, job);
+    jobMapLock_.Unlock();
+    GenScanJobs(key);
 }
 
 int ScanManager::CancelScanJob(LogicPoolID poolId, CopysetID id) {
+    ScanKey key(poolId, id);
+    // cancel scan job not started
+    {
+        WriteLockGuard writeGuard(waitSetLock_);
+        auto iter = waitScanSet_.find(key);
+        if (iter != waitScanSet_.end()) {
+            waitScanSet_.erase(iter);
+            return 0;
+        }
+    }
+    // cancel scan job started
+    auto job = GetJob(key);
+    if (nullptr != job) {
+        auto nodePtr = copysetNodeManager_->GetCopysetNode(poolId, id);
+        nodePtr->SetScan(false);
+        WriteLockGuard writeGuard(jobMapLock_);
+        jobs_.erase(key);
+    }
     return 0;
 }
 
-bool ScanManager::IsRepeatReq(LogicPoolID poolId, CopysetID id) {
-    return 0;
-}
-
-void ScanManager::StartScanJob(ScanKey key) {
-    job_.poolId = key.first;
-    job_.id = key.second;
-    job_.type = ScanType::Init;
-    job_.currentChunkId = 0;
-    job_.startTime = 0;
-    GenScanJob();
+bool ScanManager::GenScanJob(std::shared_ptr<ScanJob> job) {
+    bool done = false;
+    switch (job->type) {
+        case ScanType::Init:
+            chunkMap_ = dataStore_->GetChunkMap();
+            job->type = ScanType::NewMap;
+            break;
+        case ScanType::NewMap:
+            job->iter = chunkMap_.find(job->currentChunkId);
+            if (job->iter == chunkMap_.end()) {
+                job->iter = chunkMap_.begin();
+                job->currentChunkId = job->iter->first;
+            }
+            ScanChunkReqProcess(job);
+            done = true;
+            break;
+        case ScanType::WaitMap:
+            {
+                auto iter = job->maps.begin();
+                while (iter != job->maps.end()) {
```

Comment:
this while loop here may concurrent with `ScanChunkReqProcess`.

Reply:
> this while loop here may concurrent with `ScanChunkReqProcess`.

yes, I will add a lock to protect job->maps